### ISSUE TITLE
Dependency issue when building mc_linuxrt

### ISSUE THREAD

**Issue by magnusmusngi** [gergondet/mc_linuxrt#1]
Hi,

We are trying to test mc_linuxrt in an RT_PREEMPT kernel as suggested in https://github.com/jrl-umi3218/mc_rtc/issues/43. However, we are having dependency issues in building it. Our RT-patched kernel version is 4.14.184-rt84 and we tried adding the directory of this downloaded patched-kernel to CMakeLists.txt as an attempt to solve the problem, but it didn't worked. The original error (without adding the directory) was:

```
[ 16%] Building CXX object CMakeFiles/MCLinuxRTMainLoop.dir/src/main.cpp.o
/home/devpc/SoftwareDev/JitterInvestigation/mc_linuxrt/src/main.cpp:8:31: fatal error: linux/sched/types.h: No such file or directory
compilation terminated.
CMakeFiles/MCLinuxRTMainLoop.dir/build.make:82: recipe for target 'CMakeFiles/MCLinuxRTMainLoop.dir/src/main.cpp.o' failed
make[2]: *** [CMakeFiles/MCLinuxRTMainLoop.dir/src/main.cpp.o] Error 1
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/MCLinuxRTMainLoop.dir/all' failed
make[1]: *** [CMakeFiles/MCLinuxRTMainLoop.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

After adding the directory to the CMakeLists.txt, the new error is this:

```
[ 16%] Building CXX object CMakeFiles/MCLinuxRTMainLoop.dir/src/main.cpp.o
In file included from /home/devpc/SoftwareDev/JitterInvestigation/mc_linuxrt/src/main.cpp:5:0:
/home/devpc/Downloads/linux-4.14.184/include/linux/sched.h:12:25: fatal error: asm/current.h: No such file or directory
compilation terminated.
CMakeFiles/MCLinuxRTMainLoop.dir/build.make:82: recipe for target 'CMakeFiles/MCLinuxRTMainLoop.dir/src/main.cpp.o' failed
make[2]: *** [CMakeFiles/MCLinuxRTMainLoop.dir/src/main.cpp.o] Error 1
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/MCLinuxRTMainLoop.dir/all' failed
make[1]: *** [CMakeFiles/MCLinuxRTMainLoop.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2
```

What would you suggest to solve this issue?

Thank you

**Comment by gergondet** [comment]
Hi,

On Ubuntu and Debian this header comes from the `linux-libc-dev` package.

**Comment by magnusmusngi** [comment]
Thank you, @gergondet. I will be checking this

**Comment by magnusmusngi** [comment]
Hi @gergondet,

I installed linux-libc-dev but I'm still getting the same error. Do I need to add a specific line in the CMakeLists.txt to make this work? I also noticed that the linux-libc-dev version is 4.4.0-185.215, while our current kernel version is 4.14.184-rt84. Does these two have to match?

Thank you

**Comment by gergondet** [comment]
Hi @magnusmusngi 

> we tried adding the directory of this downloaded patched-kernel to CMakeLists.txt as an attempt to solve the problem

Did you keep this around?

**Comment by magnusmusngi** [comment]
Hi @gergondet 

I removed it from the CMakeLists.txt.

**Comment by gergondet** [comment]
So which error do you get now?

> I also noticed that the linux-libc-dev version is 4.4.0-185.215, while our current kernel version is 4.14.184-rt84. Does these two have to match?

It might be problematic but I don't have such a setup on-hand to try it out... Ideally you'd want a rt kernel that is very close to your distribution provided one if the distribution does not provide their own rt kernl

**Comment by magnusmusngi** [comment]
I still get the same error of missing "linux/sched/types.h".

> It might be problematic but I don't have such a setup on-hand to try it out... Ideally you'd want a rt kernel that is very close to your distribution provided one if the distribution does not provide their own rt kernl

I see. Our actual distribution is 4.15.0-106-generic and our RT-patched kernel is 4.14.184-rt84, and I noticed that there is no patched kernel that is 4.15. The closest is either 14.14.* or 14.16.*.

About the linux-libc-dev version of 4.4.0-185.215, sorry, I got confused when I posted that. I've mistaken it as the Linux kernel header version that should [match](https://www.tecmint.com/install-kernel-headers-in-ubuntu-and-debian/) the installed kernel version of the system (because package description says Linux Kernel Header), but actually, they are mentioning linux-headers-* package. But anyway, I tested building the code again in RT-patched kernel 4.4.225-rt198 (and also installed linux-headers-4.4.0-185) and still got the same error.

**Comment by magnusmusngi** [comment]
I just tried looking at the [file list](https://packages.ubuntu.com/xenial/amd64/linux-libc-dev/filelist) of package linux-libc-dev for my distribution, and I noticed that there is actually no /usr/include/linux/sched/types.h. The files that are closest are:

- /usr/include/asm-generic/types.h

- /usr/include/linux/iio/types.h

- /usr/include/linux/types.h

- /usr/include/x86_64-linux-gnu/asm/types.h

I tried replacing `linux/sched/types.h` with each of them, and I got a new error, but they are the same for all the aforementioned replacements:

```
[ 16%] Building CXX object CMakeFiles/MCLinuxRTMainLoop.dir/src/main.cpp.o
/home/hmr_embedded/SoftwareDev/JitterInvestigation/mc_linuxrt/src/main.cpp: In function ‘int main(int, char**)’:
/home/hmr_embedded/SoftwareDev/JitterInvestigation/mc_linuxrt/src/main.cpp:32:21: error: aggregate ‘sched_attr attr’ has incomplete type and cannot be defined
   struct sched_attr attr;
                     ^
```

**Comment by gergondet** [comment]
Hi,

You can look for which file contains the definitions for the `sched_attr` definitions.

I've built this project on Linux 4.19 (with Debian Buster) so if you have 4.15 installed it shouldn't be too far off. Maybe you'll need to indicate CMake to include your actual kernel header files and the path to the asm headers (e.g. `/usr/src/linux-headers-5.3.0-62/arch/x86/include/` but this depends on your actual architecture and you might have to pass down some extra defines to make sure the configuration matches your kernel, e.g. `CONFIG_X86_64` if you are building for x86_64)

**Comment by magnusmusngi** [comment]
Hi,

I tried building it in Ubuntu 18.04 with RT-patched kernel version 5.2.21-rt15 and it was successful. That version of Ubuntu has the file `/usr/include/linux/sched/types.h`

**Comment by magnusmusngi** [comment]
Hi,

I tried building it in Ubuntu 18.04 with RT-patched kernel version 5.2.21-rt15 and it was successful. That version of Ubuntu has the file `/usr/include/linux/sched/types.h`

**Comment by magnusmusngi** [comment]
Hi,

I tried building it in Ubuntu 18.04 with RT-patched kernel version 5.2.21-rt15 and it was successful. That version of Ubuntu has the file `/usr/include/linux/sched/types.h`

I then tried to build it again in Ubuntu 16.04 with RT-patched kernel version 4.19.127-rt55, and learned that I just [need to define sched_attr](https://www.i-programmer.info/programming/cc/13002-applying-c-deadline-scheduling.html) by including the following lines of code in `main.cpp`:

```
struct sched_attr {
    uint32_t size;
    uint32_t sched_policy;
    uint64_t sched_flags;
    int32_t sched_nice;
    uint32_t sched_priority;
    uint64_t sched_runtime;
    uint64_t sched_deadline;
    uint64_t sched_period;
};
```

It now builds successfully.

Thank you very much again for all the help.
